[WTF] Refactor EnumTraits to improve constexpr evaluation and handling of signed enums

hyjorc1 opened the pull request:
#### 9c6d843b1dc094b677f052ddd9e6c4e9f39c87cc
<pre>
[WTF] Refactor EnumTraits to improve constexpr evaluation and handling of signed enums
<a href="https://bugs.webkit.org/show_bug.cgi?id=287370">https://bugs.webkit.org/show_bug.cgi?id=287370</a>
<a href="https://rdar.apple.com/144487521">rdar://144487521</a>

Reviewed by Keith Miller.

This patch:
1. Removed unnecessary forConstexpr, replacing it with std::index_sequence for
   better compile-time performance.
2. Improved handling of signed enums by introducing enumNamesMin and enumNamesMax,
   ensuring min and max values are within the valid range of the underlying type.
3. Optimized the computation of enumNamesSize to prevent overflow and ensure
   correctness for signed and unsigned enums.
4. Updated enumName lookup logic to correctly compute indices for signed enums
   using unsigned arithmetic.
5. Added new test cases to cover signed and unsigned enums, including cases with
   negative values, large gaps, and out-of-range lookups.

This change ensures robust handling of enums with various underlying types, prevents
unexpected overflows, and enhances overall constexpr evaluation efficiency.

* Source/WTF/wtf/EnumTraits.h:
(WTF::enumName):
(WTF::enumNames): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/EnumTraits.cpp:
(TestWebKitAPI::TEST(WTF_EnumTraits, EnumNameArgument)):

Canonical link: <a href="https://commits.webkit.org/290345@main">https://commits.webkit.org/290345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08338975dd4d5a0bb84a92e8b71cd44da5c03e66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89754 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/9283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94747 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40522 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91806 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/9670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17560 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/9670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/49496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/9670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39656 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82553 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/9670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96574 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88529 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16936 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17192 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21758 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10115 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14083 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16949 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111022 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16690 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/26593 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18472 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (commented):
I think I would do this as something like:

```
template<typename E>
constexpr std::underlying_type_t<E> getEnumNamesMax()
{
    if constexpr (requires { EnumTraits<E>::max; })
        return EnumTraits<E>::max;
    else
        return 256;
}

template<typename E>
constexpr std::underlying_type_t<E> getEnumNamesMin()
{
    if constexpr (requires { EnumTraits<E>::min; })
        return EnumTraits<E>::min;
    else
        return 0;
}

template<typename E, size_t min, size_t max>
constexpr std::array<std::span<const char>, max - min> enumNames()
{
    std::array<std::span<const char>, max - min> names;

    detail::forConstexpr<min, max>([&] (auto i) {
        names[i] = enumName<static_cast<E>(static_cast<unsigned>(i))>();
    });
    return names;
}


template<typename T>
constexpr std::span<const char> enumName(T v)
{
    constexpr auto min = getEnumNamesMin();
    constexpr auto max = getEnumNamesMax();
    constexpr auto names = enumNames<std::decay_t<T>, min, max>();
    auto numericValue = static_cast<std::underlying_type_t<T>>(v);
    if (v < min || max < v)
        return { "enum out of range" };
    return names[numericValue - min];
}
```

Which has a couple notable changes from your PR:
1) You don't need to have a default version of the `EnumTraits` struct.
2) Limit and min seem weird. Let's go with max/min.
3) There's no reason to have the array hold a bunch of values between 0 and min.

danlliu reviewed (commented):
LGTM

kmiller68 commented on Source/WTF/wtf/EnumTraits.h:
```diff
@@ -215,38 +215,92 @@ constexpr std::span<const char> enumName()
+
+    if constexpr (requires { EnumTraits<E>::max; }) {
+        constexpr Underlying maxValue = static_cast<Underlying>(EnumTraits<E>::max);
+        static_assert(std::numeric_limits<Underlying>::min() <= maxValue && maxValue <= std::numeric_limits<Underlying>::max(),
```
This assert seems like it doesn't do anything because you cast it above? I think you could also maybe cover this by changing the `if` to `if constexpr (requires { { EnumTraits<E>::max } -> std::same_as<Underylying>; })`

kmiller68 commented on Source/WTF/wtf/EnumTraits.h:
```diff
@@ -215,38 +215,92 @@ constexpr std::span<const char> enumName()
+    constexpr Underlying defaultMax = std::is_signed_v<Underlying> ? INT8_MAX : UINT8_MAX;
+    constexpr Underlying computedMax = (sizeof(E) > 1) ? static_cast<Underlying>(defaultMax << 1) : defaultMax;
+
+    static_assert(std::numeric_limits<Underlying>::min() <= computedMax && computedMax <= std::numeric_limits<Underlying>::max(),
```
Ditto.

kmiller68 reviewed (approved):
r=me with a comment.

hyjorc1 committed:
[WTF] Refactor EnumTraits to improve constexpr evaluation and handling of signed enums
https://bugs.webkit.org/show_bug.cgi?id=287370
rdar://144487521

Reviewed by Keith Miller.

This patch:
1. Removed unnecessary forConstexpr, replacing it with std::index_sequence for
   better compile-time performance.
2. Improved handling of signed enums by introducing enumNamesMin and enumNamesMax,
   ensuring min and max values are within the valid range of the underlying type.
3. Optimized the computation of enumNamesSize to prevent overflow and ensure
   correctness for signed and unsigned enums.
4. Updated enumName lookup logic to correctly compute indices for signed enums
   using unsigned arithmetic.
5. Added new test cases to cover signed and unsigned enums, including cases with
   negative values, large gaps, and out-of-range lookups.

This change ensures robust handling of enums with various underlying types, prevents
unexpected overflows, and enhances overall constexpr evaluation efficiency.

* Source/WTF/wtf/EnumTraits.h:
(WTF::enumName):
(WTF::enumNames): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/EnumTraits.cpp:
(TestWebKitAPI::TEST(WTF_EnumTraits, EnumNameArgument)):

Canonical link: https://commits.webkit.org/290345@main
